### PR TITLE
Extends io traces by completion

### PR DESCRIPTION
**Pull request by mariuszbarczak**

### CODE REVIEW

**Comment by mariuszbarczak** [comment]
Fixes #51

File: source/kernel/config.h [hunk 1]
```diff
@@ -51,6 +64,49 @@ static inline int iotrace_unregister_trace_block_bio_queue(
...
+                               unsigned int nr_bytes);
+
+static inline int iotrace_register_trace_block_bio_complete(
+        iotrace_bio_complete_fn fn) {
+    int result;
+
+    result = register_trace_block_bio_complete(fn, NULL);
+    WARN_ON(result);
+    if (result) {
+        goto ERROR;
```
Comment (tomaszrybicki) [inline source/kernel/config.h]: How about REGISTER_BIO_ERROR

File: source/kernel/config.h [hunk 2]
```diff
@@ -51,6 +64,49 @@ static inline int iotrace_unregister_trace_block_bio_queue(
...
+    }
+
+    result = register_trace_block_rq_complete(iotrace_block_rq_complete, fn);
+    WARN_ON(result);
+    if (result) {
+        goto BIO_COMPLETE;
```
Comment (tomaszrybicki) [inline source/kernel/config.h]: and REGISTER_RQ_ERROR?

File: source/kernel/config.h [hunk 3]
```diff
@@ -65,6 +121,7 @@ static inline int iotrace_unregister_trace_block_bio_queue(
 /* Gets BIO vector  */
 #define IOTRACE_BIO_GET_DEV(bio) bio->bi_disk
 #define IOTRACE_LOOKUP_BDEV(path) lookup_bdev(path, 0)
+/* Check if BIO completion has been already traced */
```
Comment (tomaszrybicki) [inline source/kernel/config.h]: What does this comment pertain to?

File: source/kernel/config.h [hunk 4]
```diff
@@ -82,5 +139,56 @@ static inline int iotrace_unregister_trace_block_bio_queue(
...
+        iotrace_bio_complete_fn fn) {
+    int result;
+    char *sym_name = "__tracepoint_block_rq_complete";
+    typeof(&__tracepoint_block_rq_complete) tracepoint =
+            (void *) kallsyms_lookup_name(sym_name);
+
+    result = register_trace_block_bio_complete(fn, NULL);
+    WARN_ON(result);
+    if (result) {
+        goto ERROR;
```
Comment (tomaszrybicki) [inline source/kernel/config.h]: As in earlier comment

File: source/kernel/io_trace.c [hunk 5]
```diff
@@ -238,6 +264,34 @@ int iotrace_init_buffers(struct iotrace_context *iotrace, uint64_t size)
 	return result;
 }
 
+static int _register_trace_points(void)
+{
+    int result;
+
+    result = iotrace_register_trace_block_bio_queue(bio_queue_event);
+    if (result) {
+        goto ERROR;
```
Comment (tomaszrybicki) [inline source/kernel/io_trace.c]: As in earlier comment

File: source/kernel/config.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright(c) 2012-2018 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include "config.h"
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0)
+#define IOTRACE_BIO_TRACE_COMPLETION(bio) true
```
[the remark this answers is not in the document]
Comment (mariuszbarczak) [reply source/kernel/config.c]: I think it's ok, it indicates if trace completion of BIO.

**Comment by tomaszrybicki** [comment]
retest this please

**Comment by tomaszrybicki** [comment]
retest this please

**Comment by tomaszrybicki** [comment]
retest this please

**Comment by tomaszrybicki** [comment]
retest this please

**Pull request merged**